File saver enhancement

Jolijn2020 opened the pull request.

S010MON commented on src/main/java/app/controller/settings/Settings.java:
```diff
@@ -35,12 +36,95 @@ public void addFurniture(Rectangle2D rectangle, FurnitureType type)
+            if(obj.getType()==FurnitureType.GLASS)
+                objects.add(obj);
+        }
+        return objects;
```
This branch has inherited all of this code from #129 - still requires changes before it can be closed
Replace with something like:

```java
 public List<SettingsObject> getFurniture(FurnitureType type) 
 {
      furniture.stream()
               .filter(obj -> obj.getType() == type)
               .toList();  
 }
```

This replaces all of the get methods in one go

S010MON commented on src/main/java/app/controller/settings/SettingsObject.java:
```diff
@@ -18,16 +21,35 @@ public SettingsObject(Rectangle2D rectangle, FurnitureType type)
     }
 
-    public SettingsObject(Rectangle2D rectangle, Vector teleportTo)
+    public SettingsObject(Rectangle2D rectangle, Vector teleportTo, double teleportRotation)
```
Why are you adding this?!  We don't have the functionality to support this yet, so you are adding more complexity, more code and overloading constructors multiple times without adding any features! 

YAGNI  -You aren't gonna need it

S010MON commented on src/main/java/app/controller/settings/SettingsObject.java:
```diff
@@ -18,16 +21,35 @@ public SettingsObject(Rectangle2D rectangle, FurnitureType type)
+        this.textureOrientation = null;
+    }
+
+    public SettingsObject(Rectangle2D rectangle, Integer textureType, Integer textureOrientation)
```
YAGNI  - We don't know what a texture is, we just ignore it

S010MON commented on src/main/java/app/controller/settings/SettingsObject.java:
```diff
@@ -48,8 +70,23 @@ public String toString()
+            sb.append(" ");
+            sb.append(teleportRotation);
+        }
+        if(type == FurnitureType.TEXTURE && textureType != null)
```
YAGNI - We don't know what a texture is ...

S010MON commented on src/main/java/app/controller/settings/SettingsObject.java:
```diff
@@ -48,8 +70,23 @@ public String toString()
+            sb.append(" ");
+            sb.append(textureType);
+        }
+        if(type == FurnitureType.TEXTURE && textureOrientation != null)
```
YAGNI - We don't know what a texture is ...

S010MON commented on src/main/java/app/model/furniture/FurnitureType.java:
```diff
@@ -9,7 +9,8 @@
     INTRUDER_SPAWN("spawnAreaIntruders"),
-    TARGET("targetArea");
+    TARGET("targetArea"),
+    TEXTURE("texture");
```
YAGNI - We don't know what a texture is ...

S010MON reviewed (changes requested):
Look,  I understand you are just trying to be thorough, but you are adding loads of complexity to reading in these files, but you aren't adding features.  We don't even know what a texture is, but you've added a load of code to the codebase (and you have not dealt with the dependency issues of adding extra enums without taking care of the switch cases that rely on them) that doesn't do anything but make the objects bigger.

Work on these issues isn't wrong, but work on them start to finish:

1. Add a rotation to the actual objects
2. Realise that there is not a way to read it in
3. Parse the file for rotation as well
4. Realise that you have to refactor the writing
5. Change the writing

If you go through all of them, then you not only deal with all dependencies start to finish, you implement a full feature in one go.  Additionally you don't write any code that isn't used.  (Think all the getters you wrote ... they only get called by your tests)

I'm not trying to be a dick about this, I'm just trying to improve your development flow so that you contribute useful LoC rather than ones that don't get run

Jolijn2020 commented:
I understand what you mean about not adding in things we don't use, but the examiners confirmed that the map_file we got contained all the minimal features we have to implement. So we will have to use all of them at some point.

So I would rather do all of them at once instead of having to keep adding them each time we start working with it. Especially since if we forgot about it (or someone else starts working with it who doesn't know) it could lead to some hard to locate bugs.

I have also already done most of the things you mentioned.

1. Add a rotation to the actual objects
          I did that in SettingsObject, also added a get method for it as well

2. Realise that there is not a way to read it in
3. Parse the file for rotation as well
            I realized that, and added it to the FileParser (loadSetting) as well

4. Realise that you have to refactor the writing
            I am not 100% sure what you mean with that
            The writing of all the furniture uses the SettingsObject toString method, I adjusted that to work for teleport and texture without impacting the other furnitures.
            And I'm not sure what else would need to be refactered.

5. Change the writing
            I am not 100% sure what you mean with that
            The writing of all the furniture uses the SettingsObject toString method, I adjusted that to work for teleport and texture without impacting the other furnitures.

6. Finish the feature
            I think I that? I made sure all uses of teleport were adjusted.
            And I changed FileParser (loadSetting) to reflect the changes of teleport and added texture.

S010MON commented:
> Add a rotation to the actual objects
> I did that in SettingsObject, also added a get method for it as well
> And as far as I was aware, teleport hasn't been implemented yet, so couldn't add the rotation in the agent itself yet.

This is exactly what I mean, you added rotation to a settings object.  Does that get created in a factory?  Does that get rotated on the map?  Does that get rendered? Does that get painted?

Because you've only worked on the settings/settingsObject you aren't looking down the line at the rest of the code that is effected.  I'm just trying to get you more involved in the actual important features, rather than just making the objects full of unused variables.

Jolijn2020 commented:
> This is exactly what I mean, you added rotation to a settings object. Does that get created in a factory? Does that get rotated on the map? Does that get rendered? Does that get painted?

I added it to every creation of teleport already there as 0.0, so if got created in a factory, it's already adjusted.

The rotation of the agent is something that doesn't come into play until teleportation itself is implemented (the issue for that is still unclaimed, so shouldn't be yet). It simply adjusts the angle the agent is looking in. This is something that doesn't need a seperate animation, it will reflect in the animation of the agent itself (it's angle). This is something that is needed to implement the teleport as it determines the new angle of the agent after teleportation. (unless you just ignore this part of functionality, but you shouldn't)

I will implement the rotation today/tomorrow and add in the rotation though.


As for the textures, I did not add that to factory indeed.
The FurnitureFactory doesn't seem to contain every single type of furniture, so didn't think it would be a good Idea to just randomly add texture (or teleport).

I looked through the files and did find a list of all furnitures in FurniturePane though, so will add it there. That is the rendering you are talking about I assume.
